rclcpp/include/rclcpp/any_subscription_callback.hpp

```diff
@@ -15,256 +15,338 @@
-#include "tracetools/utils.hpp"
+
+template<class> inline constexpr bool always_false_v = false;
 
 namespace rclcpp
 {
 
-template<typename MessageT, typename Alloc>
-class AnySubscriptionCallback
+namespace detail
+{
+
+template<typename MessageT, typename AllocatorT>
+struct MessageDeleterHelper
 {
-  using MessageAllocTraits = allocator::AllocRebind<MessageT, Alloc>;
+  using MessageAllocTraits = allocator::AllocRebind<MessageT, AllocatorT>;
   using MessageAlloc = typename MessageAllocTraits::allocator_type;
   using MessageDeleter = allocator::Deleter<MessageAlloc, MessageT>;
-  using ConstMessageSharedPtr = std::shared_ptr<const MessageT>;
-  using MessageUniquePtr = std::unique_ptr<MessageT, MessageDeleter>;
+};
 
-  using SharedPtrCallback = std::function<void (const std::shared_ptr<MessageT>)>;
-  using SharedPtrWithInfoCallback =
-    std::function<void (const std::shared_ptr<MessageT>, const rclcpp::MessageInfo &)>;
-  using ConstSharedPtrCallback = std::function<void (const std::shared_ptr<const MessageT>)>;
-  using ConstSharedPtrWithInfoCallback =
-    std::function<void (const std::shared_ptr<const MessageT>, const rclcpp::MessageInfo &)>;
-  using UniquePtrCallback = std::function<void (MessageUniquePtr)>;
+template<typename MessageT, typename AllocatorT>
+struct AnySubscriptionCallbackHelper
+{
+  using MessageDeleter = typename MessageDeleterHelper<MessageT, AllocatorT>::MessageDeleter;
+
+  using ConstRefCallback =
+    std::function<void (const MessageT &)>;
+  using ConstRefWithInfoCallback =
+    std::function<void (const MessageT &, const rclcpp::MessageInfo &)>;
+
+  using UniquePtrCallback =
+    std::function<void (std::unique_ptr<MessageT, MessageDeleter>)>;
   using UniquePtrWithInfoCallback =
-    std::function<void (MessageUniquePtr, const rclcpp::MessageInfo &)>;
+    std::function<void (std::unique_ptr<MessageT, MessageDeleter>, const rclcpp::MessageInfo &)>;
+
+  // TODO(wjwwood): should we drop this one, because you can achieve it easily
+  //   with the const shared_ptr & version?
+  using SharedConstPtrCallback =
+    std::function<void (std::shared_ptr<const MessageT>)>;
+  using SharedConstPtrWithInfoCallback =
+    std::function<void (std::shared_ptr<const MessageT>, const rclcpp::MessageInfo &)>;
+
+  // TODO(wjwwood): should we drop this one, because you should prefer the
+  //   shared_ptr one (if you intend to actuall share ownership) or the
+  //   const MessageT & one if you just want to read the message?
+  using ConstRefSharedConstPtrCallback =
+    std::function<void (const std::shared_ptr<const MessageT> &)>;
+  using ConstRefSharedConstPtrWithInfoCallback =
+    std::function<void (const std::shared_ptr<const MessageT> &, const rclcpp::MessageInfo &)>;
+
+  // Deprecated signatures:
+  using SharedPtrCallback =
+    std::function<void (std::shared_ptr<MessageT>)>;
+  using SharedPtrWithInfoCallback =
+    std::function<void (std::shared_ptr<MessageT>, const rclcpp::MessageInfo &)>;
```

Comment:
RFC

Reply:
:+1:

Reply:
:+1:  It'll be disruptive, but sounds like a safer direction to move in.

Reply:
We could keep the signature and just make a copy for each subscription that asks for this, but I think it adds to the clutter and just isn't necessary. I'd prefer to put the allocation of the copy onto the user so they can control it. Making those copies would cause us to do the allocation and it might not be trivial for the user to influence how that happens.

Reply:
> 👍 It'll be disruptive, but sounds like a safer direction to move in.

It will be disruptive, I think quite a lot of people are using this right now. I haven't addressed all the deprecation warnings yet just because I wanted to see how we landed on this signature.

Luckily we've had the `shared_ptr<const MessageT>` for a while so people can use that in older versions if they have one branch for several distributions. Otherwise, I'd say we should consider backporting some new alternatives to this one.

Reply:
I'm running a CI job (linux only, only cyclone) to see what deprecation warnings come up:

[![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux&build=14123)](https://ci.ros2.org/job/ci_linux/14123/)

Reply:
So it failed because we don't have a way to propagate the C++17 standard being selected in cmake. I'm working on a way to address that, but you could browse the warnings from `rclcpp` which built.

Reply:
Though I would have preferred to push the deprecation forward, I'm going to drop the deprecation from this pull request, due to concerns about how close it is to the Galactic API freeze made by other maintainers.

Reply:
We decided not to deprecate this until after galactic, but I will be discouraging its use in the doc updates I am going to make, and I will "fix" all the places it would have been a warning that I can. I already did it for this repository.